docs/internals/writing-a-provider.rst

```diff
@@ -0,0 +1,134 @@
+the context is available for use, it initializes **options** and **services**. We will example both of
+these to understand what they represent and how EF configures itself to use
+different providers.
+
+Options
+^^^^^^^
+
+``Microsoft.EntityFrameworkCore.Infrastructure.DbContextOptions`` is the API surface for **users** to configure ``DbContext``. Provider
+writers are responsible for creating API to configure options and to make
+services responsive to these options. For example, most providers require a
+connection string. These options are
+typically created using ``DbContextOptionsBuilder``.
+
+Services
+^^^^^^^^
+
+``System.IServiceProvider`` is the main interface used for interaction with services.
+EF makes heavy use of `dependency injection (DI)
+<https://wikipedia.org/wiki/Dependency_injection>`_. The ``ServiceProvider``
+contains a collection of services available for injection. Initialization uses
+``DbContextOptions`` to add additional services if needed and  select
+a scoped set of services that all EF operations will use during execution.
+
+See also :doc:`services`.
+
+.. note::
+  EF uses `Microsoft.Extensions.DependencyInjection <https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection/>`_ to implement dependency injection. Documentation for this project `is available on docs.asp.net <https://docs.asp.net/en/latest/fundamentals/dependency-injection.html>`_.
+
+Plugging in a Provider
+----------------------
+
+As explained above, EF uses options and services. Each provider must create API
+so users to add provider-specific options and services. This API is best created
+by using extension methods.
+
+.. tip::
+  When defining an extension method, define it in the namespace of the object
+  being extended so Visual Studio auto-complete will include the extension
+  method as a possible completion.
+
+The `Use` Method
+^^^^^^^^^^^^^^^^
+
+By convention, providers define a ``UseX()`` extension on ``DbContextOptionsBuilder``. This configures **options** which it typically takes as arguments to method.
+
+::
+
+  optionsBuilder.UseMyProvider("Server=contoso.com")
+
+The ``UseX()`` extension method creates a provider-specific implementation of
+``IDbContextOptionsExtension`` which is added to the collection of extensions
+stored within ``DbContextOptions``. This is done by calling a hidden API call to
+``IDbContextOptionsBuilderInfrastructure.AddOrUpdateExtension``.
+
+.. literalinclude:: sample/EntityFrameworkCore.ProviderStarter/Extensions/MyProviderDbContextOptionsExtensions.cs
+  :language: csharp
+  :caption: An example implementation of the "Use" method
+  :linenos:
+  :lines: 8-18
+
+.. tip::
+  The ``UseX()`` method can also be used to return a special wrapper around
+  ``DbContextOptionsBuilder`` that allows users to configure multiple options
+  with change calls. See ``SqlServerDbContextOptionsBuilder`` as an example.
+
+The `Add` Method
+^^^^^^^^^^^^^^^^
+
```

Comment:
We should provide a list of all the services that need to be implemented (or tell them how to derive it from the code). I think this is as simple as needing to implement everything that is exposed from `DatabaseProviderServices`? Though we should cover that those things should be resolved from DI and not just explicitly returned from an implementation of `DatabaseProviderServices`. 

We should also explain what `DatabaseProviderServices` is somewhere in this doc, since it is pretty fundamental to how things work.